node/src/Pos/Block/Network/Logic.hs

```diff
@@ -259,36 +262,47 @@ matchRequestedHeaders headers mgh@MsgGetHeaders {..} inRecovery =
           | otherwise -> MRGood
 
 requestHeaders
-    :: forall ssc ctx m t.
+    :: forall ssc ctx m.
        (SscWorkersClass ssc, WorkMode ssc ctx m)
-    => (NewestFirst NE (BlockHeader ssc) -> m t)
+    => (NewestFirst NE (BlockHeader ssc) -> m ())
     -> MsgGetHeaders
     -> NodeId
     -> ConversationActions MsgGetHeaders (MsgHeaders ssc) m
-    -> m (Maybe t)
+    -> m ()
 requestHeaders cont mgh nodeId conv = do
     logDebug $ sformat ("requestHeaders: sending "%build) mgh
     send conv mgh
     mHeaders <- recvLimited conv
     inRecovery <- recoveryInProgress
+    -- TODO: it's very suspicious to see False here as requestHeaders
+    -- is only called when we're in recovery mode.
     logDebug $ sformat ("requestHeaders: inRecovery = "%shown) inRecovery
-    flip (maybe onNothing) mHeaders $ \(MsgHeaders headers) -> do
-        logDebug $ sformat
-            ("requestHeaders: received "%int%" headers of total size "%builder%
-             " from nodeId "%build%": "%listJson)
-            (headers ^. _Wrapped . to NE.length)
-            (unitBuilder $ biSize headers)
-            nodeId
-            (map headerHash headers)
-        case matchRequestedHeaders headers mgh inRecovery of
-            MRGood           -> do
-                handleRequestedHeaders cont inRecovery headers
-            MRUnexpected msg -> handleUnexpected headers msg
+    case mHeaders of
+        Nothing -> do
+            logWarning "requestHeaders: received Nothing as a response on MsgGetHeaders"
+            throwM $ DialogUnexpected $
+                sformat ("requestHeaders: received Nothing from "%build) nodeId
+        Just (MsgNoHeaders t) -> do
+            logWarning $ "requestHeaders: received MsgNoHeaders: " <> t
+            throwM $ DialogUnexpected $
```

Comment:
Is this really unexpected/exceptional? I guess this request is only made if we've somehow discovered that this peer really does have the relevant headers. Still we might not want to throw, because it will appear as a failure to the outbound queue when in fact the peer is completely fine.

Reply:
I do not understand your model. We either succeed to communicate with peer or not. If we have any problems with peer, we ignore him for a while. Recovery worker is written using `m ()` actions and every failure is represented by the exception. Exception handler in recovery worker kicks user out and finds new recovery server. Every failure to retrieve/process headers is an exception.